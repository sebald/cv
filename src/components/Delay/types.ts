import * as React from 'react';

export interface DelayState {
  appear: boolean;
}

export interface DelayProps {
  timeout: number;
  render: (() => React.ReactNode);
}
