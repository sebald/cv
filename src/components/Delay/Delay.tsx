import * as React from 'react';

import { DelayProps, DelayState } from './types';

export class Delay extends React.Component<DelayProps, DelayState> {
  state: DelayState = {
    appear: false,
  };

  timer: number | null = null;

  clearTimer() {
    if (this.timer !== null) {
      clearTimeout(this.timer);
    }
  }

  update() {
    const { timeout } = this.props;
    this.clearTimer();

    /**
     * Use `window` object so TypeScript can infer the correct type.
     * If omitted it would use node's `setTimeout` signature.
     */
    this.timer = window.setTimeout(() => {
      this.setState({
        appear: true,
      });
    }, timeout);
  }

  componentDidMount() {
    this.update();
  }

  componentWillUnmount() {
    this.clearTimer();
  }

  render() {
    return this.state.appear ? this.props.render() : null;
  }
}
