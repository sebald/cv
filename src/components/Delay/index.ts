export * from './Delay';
export * from './types';
