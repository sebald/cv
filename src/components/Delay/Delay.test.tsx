import { mount } from 'enzyme';
import * as React from 'react';

import { Delay } from './Delay';

jest.useFakeTimers();

test('correctly render component', () => {
  const wrapper = mount(<Delay timeout={100} render={() => <div />} />);
  expect(wrapper).toMatchSnapshot('not rendered');

  jest.runAllTimers();
  expect(wrapper).toMatchSnapshot('rendered');
});

test('appear after specifies timeout', () => {
  const spy = jest.fn(() => <div />);
  mount(<Delay timeout={100} render={spy} />);

  expect(spy).not.toHaveBeenCalled();
  jest.runTimersToTime(100);
  expect(spy).toHaveBeenCalled();
});
