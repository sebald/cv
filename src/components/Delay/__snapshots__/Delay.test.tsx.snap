// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`correctly render component: not rendered 1`] = `
<Delay
  render={[Function]}
  timeout={100}
/>
`;

exports[`correctly render component: rendered 1`] = `
<Delay
  render={[Function]}
  timeout={100}
/>
`;
